File: sdk-nodejs/keygen.js
const crypto = require('crypto');

/**
 * DIDKeyPair represents a generated DID with its associated key material.
 * @typedef {Object} DIDKeyPair
 * @property {string} did - The did:jwk identifier
 * @property {string} algorithm - EdDSA or ES256
 * @property {string} publicJWK - Public key in JWK format (JSON string)
 * @property {string} privateKeyPEM - Private key in PEM format
 */

/**
 * Generate a new DID:JWK with embedded public key.
 * 
 * @param {string} algorithm - Signing algorithm: 'EdDSA' (Ed25519) or 'ES256' (ECDSA P-256)
 * @returns {Promise<DIDKeyPair>} The generated DID and key material
 * 
 * @example
 * const { generateDIDJWK } = require('./keygen');
 * 
 * // Generate EdDSA key
 * const keypair = await generateDIDJWK('EdDSA');
 * console.log('DID:', keypair.did);
 * 
 * // Save private key to file if needed
 * const fs = require('fs');
 * fs.writeFileSync('key.pem', keypair.privateKeyPEM, { mode: 0o600 });
 */
async function generateDIDJWK(algorithm = 'EdDSA') {
  if (algorithm === 'EdDSA') {
    return generateEd25519DID();
  } else if (algorithm === 'ES256') {
    return generateES256DID();
  } else {
    throw new Error(`Unsupported algorithm: ${algorithm} (supported: EdDSA, ES256)`);
  }
}

/**
 * Generate an Ed25519 key pair and format as did:jwk.
 * @private
 */
function generateEd25519DID() {
  return new Promise((resolve, reject) => {
    try {
      // Generate Ed25519 key pair
      const { publicKey, privateKey } = crypto.generateKeyPairSync('ed25519', {
        publicKeyEncoding: {
          type: 'spki',
          format: 'der'
        },
        privateKeyEncoding: {
          type: 'pkcs8',
          format: 'pem'
        }
      });

      // Extract raw public key (last 32 bytes of SPKI format)
      const rawPublicKey = publicKey.slice(-32);

      // Create public JWK
      const publicJWK = {
        kty: 'OKP',
        crv: 'Ed25519',
        x: base64urlEncode(rawPublicKey)
      };

      const publicJWKString = JSON.stringify(publicJWK);

      // Create DID by base64url-encoding the public JWK
      const did = 'did:jwk:' + base64urlEncode(Buffer.from(publicJWKString));

      resolve({
        did,
        algorithm: 'EdDSA',
        publicJWK: publicJWKString,
        privateKeyPEM: privateKey
      });
    } catch (error) {
      reject(new Error(`Failed to generate Ed25519 key: ${error.message}`));
    }
  });
}

/**
 * Generate an ECDSA P-256 key pair and format as did:jwk.
 * @private
 */
function generateES256DID() {
  return new Promise((resolve, reject) => {
    try {
      // Generate ECDSA P-256 key pair
      const { publicKey, privateKey } = crypto.generateKeyPairSync('ec', {
        namedCurve: 'prime256v1', // P-256
        publicKeyEncoding: {
          type: 'spki',
          format: 'der'
        },
        privateKeyEncoding: {
          type: 'pkcs8',
          format: 'pem'
        }
      });

      // Parse the DER-encoded public key to extract x and y coordinates
      // ECDSA P-256 public key in uncompressed format: 0x04 || x || y
      // The uncompressed point is in the last 65 bytes of the SPKI structure
      const publicKeyObject = crypto.createPublicKey({
        key: publicKey,
        format: 'der',
        type: 'spki'
      });

      // Export as JWK to get x and y
      const jwk = publicKeyObject.export({ format: 'jwk' });

      // Create public JWK
      const publicJWK = {
        kty: 'EC',
        crv: 'P-256',
        x: jwk.x,
        y: jwk.y
      };

      const publicJWKString = JSON.stringify(publicJWK);

      // Create DID by base64url-encoding the public JWK
      const did = 'did:jwk:' + base64urlEncode(Buffer.from(publicJWKString));

      resolve({
        did,
        algorithm: 'ES256',
        publicJWK: publicJWKString,
        privateKeyPEM: privateKey
      });
    } catch (error) {
      reject(new Error(`Failed to generate ES256 key: ${error.message}`));
    }
  });
}

/**
 * Base64url encode a buffer without padding.
 * @private
 */
function base64urlEncode(buffer) {
  return buffer.toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}

module.exports = {
  generateDIDJWK
};

File: sdk-go/keygen.go
package sdk

import (
	"crypto"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"fmt"
)

// DIDKeyPair represents a generated DID with its associated key material.
type DIDKeyPair struct {
	DID           string        // The did:jwk identifier
	Algorithm     string        // EdDSA or ES256
	PublicJWK     string        // Public key in JWK format (JSON)
	PrivateKeyPEM string        // Private key in PEM format (PKCS8)
	PrivateKey    crypto.Signer // The actual private key for signing
}

// GenerateDIDJWK generates a new key pair and returns a did:jwk identifier
// along with the key material. Supported algorithms: "EdDSA" (Ed25519) and "ES256" (ECDSA P-256).
//
// Example:
//
//	keypair, err := sdk.GenerateDIDJWK("EdDSA")
//	if err != nil {
//	    log.Fatal(err)
//	}
//	fmt.Println("DID:", keypair.DID)
//	// Save private key to file if needed
//	err = os.WriteFile("key.pem", []byte(keypair.PrivateKeyPEM), 0600)
func GenerateDIDJWK(algorithm string) (*DIDKeyPair, error) {
	switch algorithm {
	case "EdDSA":
		return generateEd25519DID()
	case "ES256":
		return generateES256DID()
	default:
		return nil, fmt.Errorf("unsupported algorithm: %s (supported: EdDSA, ES256)", algorithm)
	}
}

// generateEd25519DID generates an Ed25519 key pair and formats it as did:jwk.
func generateEd25519DID() (*DIDKeyPair, error) {
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("generate ed25519 key: %w", err)
	}

	// Create public JWK
	publicJWK := map[string]string{
		"kty": "OKP",
		"crv": "Ed25519",
		"x":   base64.RawURLEncoding.EncodeToString(pub),
	}

	publicJWKBytes, err := json.Marshal(publicJWK)
	if err != nil {
		return nil, fmt.Errorf("marshal public jwk: %w", err)
	}

	// Create DID by base64url-encoding the public JWK
	did := "did:jwk:" + base64.RawURLEncoding.EncodeToString(publicJWKBytes)

	// Convert private key to PKCS8 PEM format
	privPKCS8, err := x509.MarshalPKCS8PrivateKey(priv)
	if err != nil {
		return nil, fmt.Errorf("marshal private key: %w", err)
	}

	privPEM := pem.EncodeToMemory(&pem.Block{
		Type:  "PRIVATE KEY",
		Bytes: privPKCS8,
	})

	return &DIDKeyPair{
		DID:           did,
		Algorithm:     "EdDSA",
		PublicJWK:     string(publicJWKBytes),
		PrivateKeyPEM: string(privPEM),
		PrivateKey:    priv,
	}, nil
}

// generateES256DID generates an ECDSA P-256 key pair and formats it as did:jwk.
func generateES256DID() (*DIDKeyPair, error) {
	priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("generate ecdsa key: %w", err)
	}

	// Create public JWK
	publicJWK := map[string]string{
		"kty": "EC",
		"crv": "P-256",
		"x":   base64.RawURLEncoding.EncodeToString(priv.X.Bytes()),
		"y":   base64.RawURLEncoding.EncodeToString(priv.Y.Bytes()),
	}

	publicJWKBytes, err := json.Marshal(publicJWK)
	if err != nil {
		return nil, fmt.Errorf("marshal public jwk: %w", err)
	}

	// Create DID by base64url-encoding the public JWK
	did := "did:jwk:" + base64.RawURLEncoding.EncodeToString(publicJWKBytes)

	// Convert private key to PKCS8 PEM format
	privPKCS8, err := x509.MarshalPKCS8PrivateKey(priv)
	if err != nil {
		return nil, fmt.Errorf("marshal private key: %w", err)
	}

	privPEM := pem.EncodeToMemory(&pem.Block{
		Type:  "PRIVATE KEY",
		Bytes: privPKCS8,
	})

	return &DIDKeyPair{
		DID:           did,
		Algorithm:     "ES256",
		PublicJWK:     string(publicJWKBytes),
		PrivateKeyPEM: string(privPEM),
		PrivateKey:    priv,
	}, nil
}
